File: packages/observability/logging.py
from __future__ import annotations

import json
import logging
import sys
from datetime import UTC, datetime
from typing import Any


def configure_logging(level: str = "INFO") -> None:
    logger = logging.getLogger("multi_model_rag")
    if logger.handlers:
        logger.setLevel(level.upper())
        return

    handler = logging.StreamHandler(sys.stdout)
    handler.setFormatter(logging.Formatter("%(message)s"))
    logger.addHandler(handler)
    logger.setLevel(level.upper())
    logger.propagate = False


def get_logger() -> logging.Logger:
    return logging.getLogger("multi_model_rag")


def log_event(event: str, **fields: Any) -> None:
    payload: dict[str, Any] = {
        "timestamp": datetime.now(UTC).isoformat(),
        "event": event,
        **fields,
    }
    get_logger().info(json.dumps(payload, default=str))

File: internal/observability/logging.go
package observability

import (
	"encoding/json"
	"log"
	"os"
	"strings"
	"time"
)

func ConfigureLogging(level string) {
	log.SetOutput(os.Stdout)
	log.SetFlags(0)
	LogEvent("logging.configured", map[string]any{
		"level": strings.ToUpper(strings.TrimSpace(level)),
	})
}

func LogEvent(event string, fields map[string]any) {
	payload := map[string]any{
		"timestamp": time.Now().UTC().Format(time.RFC3339Nano),
		"event":     event,
	}
	for key, value := range fields {
		payload[key] = value
	}

	data, err := json.Marshal(payload)
	if err != nil {
		log.Printf(`{"timestamp":"%s","event":"logging.error","detail":"marshal failed","original_event":%q}`, time.Now().UTC().Format(time.RFC3339Nano), event)
		return
	}
	log.Print(string(data))
}
